### PR TITLE
Seq2seq

### PR DESCRIPTION
**Pull request by JerryYanWan**
Add tests for LabeledSentence, support array label and single label
modify rnn test to select the output with largest probability each recurrent time

### CODE REVIEW

File: dl/src/main/scala/com/intel/analytics/bigdl/dataset/text/LabeledSentenceToSample.scala [hunk 1]
```diff
@@ -60,24 +58,40 @@ class LabeledSentenceToSample(vocabLength: Int,
       // initialize featureBuffer to 0.0
 
       util.Arrays.fill(featureBuffer, 0, featureBuffer.length, 0.0f)
+      util.Arrays.fill(labelBuffer, 0, labelBuffer.length, 0.0f)
 
       // One-Hot format for feature
+      // Expected transformed format should be:
```
Comment (JerryYanWan) [inline dl/src/main/scala/com/intel/analytics/bigdl/dataset/text/LabeledSentenceToSample.scala]: fixed

**Comment by JerryYanWan** [comment]
Hi, I have split the text data processing part. DocumentTokenizer.scala is a transformer that will tokenize all the sentences from disk file. (a stanford-parser tokenizer) The other people can use this transformer too. A Dictionary.scala that can either create dictionary from Array[Array[String]] (namely tokenized sentences) or from disk files (dictionary.txt that has already been created.)

File: dl/src/main/scala/com/intel/analytics/bigdl/dataset/text/Dictionary.scala [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+     _word2index.foreach(x =>
+       logger.info(x._1 + " -> " + x._2))
+   }
+
+   def printDiscard(): Unit = {
+     _disvardVocab.foreach(x =>
+       logger.info(x))
+   }
+
+   def this(sentences: Array[Array[String]],
```
Comment (psyyz10) [inline dl/src/main/scala/com/intel/analytics/bigdl/dataset/text/Dictionary.scala]: How about RDD[Array[String]] ?

File: dl/src/main/scala/com/intel/analytics/bigdl/dataset/text/Dictionary.scala [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+     _word2index = _vocabulary.zipWithIndex.toMap
+     _index2word = _word2index.map(x => (x._2, x._1))
+     _disvardVocab = freqDict.take(freqDict.length - length).map(_._1)
+     _discardSize = _disvardVocab.length
+   }
+
+   def this(directory: String) = {
+     this()
+
+     val dictionaryFile = new File(directory, "dictionary.txt")
```
Comment (psyyz10) [inline dl/src/main/scala/com/intel/analytics/bigdl/dataset/text/Dictionary.scala]: Maybe we need to read file from HDFS or with spark `sc.textFile("data.txt")` ?

File: dl/src/main/scala/com/intel/analytics/bigdl/dataset/text/Dictionary.scala [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+   private var _discardSize: Int = 0
+   private var _word2index: Map[String, Int] = null
+   private var _index2word: Map[Int, String] = null
+   private var _vocabulary: Seq[String] = null
+   private var _disvardVocab: Seq[String] = null
+}
+
+object Dictionary {
+  def apply(sentences: Array[Array[String]],
+            vocabSize: Int)
```
Comment (psyyz10) [inline dl/src/main/scala/com/intel/analytics/bigdl/dataset/text/Dictionary.scala]: It is better to set vocabSize a default value, which means the user do not want to add any restrictions to the size of vocabulary?

File: dl/src/main/scala/com/intel/analytics/bigdl/dataset/text/Dictionary.scala [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+     _discardSize = _disvardVocab.length
+   }
+
+   val logger = Logger.getLogger(getClass)
+   private var _vocabSize: Int = 0
+   private var _discardSize: Int = 0
+   private var _word2index: Map[String, Int] = null
+   private var _index2word: Map[Int, String] = null
+   private var _vocabulary: Seq[String] = null
+   private var _disvardVocab: Seq[String] = null
```
Comment (psyyz10) [inline dl/src/main/scala/com/intel/analytics/bigdl/dataset/text/Dictionary.scala]: Is it better to name `_discardVocab`?

File: dl/src/main/scala/com/intel/analytics/bigdl/dataset/text/Dictionary.scala [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+       .toMap[String, Int]
+     _index2word = _word2index.map(x => (x._2, x._1))
+     _vocabulary = _word2index.keys.toSeq
+     _vocabSize = _word2index.size
+     _disvardVocab = Source.fromFile(discardFile.getAbsolutePath)
+       .getLines().toSeq
+     _discardSize = _disvardVocab.length
+   }
+
+   val logger = Logger.getLogger(getClass)
```
Comment (psyyz10) [inline dl/src/main/scala/com/intel/analytics/bigdl/dataset/text/Dictionary.scala]: In my opinion, we offten need the frequency of the words, is it a bad idea to add frequencies as a member here as well?

File: dl/src/main/scala/com/intel/analytics/bigdl/dataset/text/Dictionary.scala (hunk 4 above)
Comment (psyyz10) [inline dl/src/main/scala/com/intel/analytics/bigdl/dataset/text/Dictionary.scala]: Is it better to add arguments such as 'needDiscard: Boolean = true' to set whether the user need discardVocab? I'm not sure :-)

File: dl/src/main/scala/com/intel/analytics/bigdl/dataset/text/Dictionary.scala [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+   }
+
+   def printDiscard(): Unit = {
+     _disvardVocab.foreach(x =>
+       logger.info(x))
+   }
+
+   def this(sentences: Array[Array[String]],
+            vocabSize: Int) = {
+     this()
```
Comment (psyyz10) [inline dl/src/main/scala/com/intel/analytics/bigdl/dataset/text/Dictionary.scala]: How about use some RDD transformations? which could be much faster

File: dl/src/main/scala/com/intel/analytics/bigdl/dataset/text/Dictionary.scala [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+     val freqDict = sentences
+       .flatMap(x => x)
+       .foldLeft(Map.empty[String, Int]) {
+         (count, word) => count + (word -> (count.getOrElse(word, 0) + 1))
+       }.toSeq.sortBy(_._2)
+
```
Comment (psyyz10) [inline dl/src/main/scala/com/intel/analytics/bigdl/dataset/text/Dictionary.scala]: How about use some RDD transformations? which could be much faster

**Comment by yiheng** [comment]
Don't commit other's change

**Pull request closed without merging**